File: 2024/13/13.rb
#!/usr/bin/env ruby

require 'matrix'
V = Vector
A_TOKENS = 3
B_TOKENS = 1
PART2_DELTA = 10_000_000_000_000

def calculate_cost(m, v)
  a, b = *m.lup.solve(v).to_a # matrix math!
  # Accept only non-negative integer solutions
  return 0 unless a.denominator == 1 and b.denominator == 1 and a >= 0 and b >= 0
  (a.to_i * A_TOKENS + b.to_i * B_TOKENS)
end

games = []
until $stdin.eof?
  x1, y1 = *$stdin.readline.split(': ').last.split(', ').map{ _1[1..].to_i }
  x2, y2 = *$stdin.readline.split(': ').last.split(', ').map{ _1[1..].to_i }
  prize = V[*$stdin.readline.split(': ').last.split(', ').map{ _1[2..].to_i }]
  games << [Matrix[[x1, x2], [y1, y2]], prize]
  $stdin.readline unless $stdin.eof? # burn empty line between games
end

sum1, sum2 = [0, 0]
games.each do |game|
  m, b = *game
  sum1 += calculate_cost(m, b)
  sum2 += calculate_cost(m, V[b[0]+PART2_DELTA, b[1]+PART2_DELTA])
end

puts sum1, sum2

File: 2024/13/13.go
package main

import (
	"bufio"
	"image"
	"os"
	"strconv"
	"strings"
)

type Game struct {
	a image.Point
	b image.Point
	p image.Point
}

var CostA int = 3
var CostB int = 1
var Part2Delta int = 10000000000000

func atoi(s string) int {
	i, _ := strconv.Atoi(s)
	return i
}

func readNumbers(line string) (int, int) {
	parts := strings.Split(strings.Split(line, ": ")[1], ", ")
	return atoi(parts[0][2:]), atoi(parts[1][2:])
}

func parseInput() []Game {
	scanner := bufio.NewScanner(os.Stdin)
	games := []Game{}
	for scanner.Scan() {
		x1, y1 := readNumbers(scanner.Text())
		scanner.Scan()
		x2, y2 := readNumbers(scanner.Text())
		scanner.Scan()
		p1, p2 := readNumbers(scanner.Text())
		if scanner.Scan() {
			scanner.Text() // burn empty line
		}
		games = append(games, Game{image.Point{x1, y1}, image.Point{x2, y2}, image.Point{p1, p2}})
	}
	return games
}

func calculateCost(game *Game) int {
	a, b, p := game.a, game.b, game.p
	ad := (b.Y*p.X - b.X*p.Y) / (a.X*b.Y - a.Y*b.X)
	bd := (a.Y*p.X - a.X*p.Y) / (a.Y*b.X - a.X*b.Y)
	if a.Mul(ad).Add(b.Mul(bd)) == p {
		return ad*CostA + bd*CostB
	}
	return 0
}

func main() {
	games := parseInput()

	var sum1, sum2 int = 0, 0
	for _, game := range games {
		sum1 += calculateCost(&game)
		game.p.X += Part2Delta
		game.p.Y += Part2Delta
		sum2 += calculateCost(&game)
	}
	println(sum1)
	println(sum2)
}
